csharp/test/Microsoft.ML.OnnxRuntime.Tests.Common/InferenceTest.cs

```diff
@@ -93,18 +93,34 @@ public void TestSessionOptions()
                 opt.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_EXTENDED;
                 Assert.Equal(GraphOptimizationLevel.ORT_ENABLE_EXTENDED, opt.GraphOptimizationLevel);
 
-                Assert.Throws<OnnxRuntimeException>(() => { opt.GraphOptimizationLevel = (GraphOptimizationLevel)10; });
+                AssertUtils.AssertThrowsCheckException<OnnxRuntimeException>(
+                    () => { opt.GraphOptimizationLevel = (GraphOptimizationLevel)10; },
+                    "Set an invalid Graph Optimization Level.");
 
                 opt.AddSessionConfigEntry("key", "value");
 
-                var ex = Assert.Throws<OnnxRuntimeException>(() => { opt.AddSessionConfigEntry("", "invalid key"); });
-                Assert.Contains("[ErrorCode:InvalidArgument] Config key is empty", ex.Message);
+                AssertUtils.AssertThrowsCheckException<OnnxRuntimeException>(
+                    () => { opt.AddSessionConfigEntry("", "invalid key"); },
+                    "Added an invalid config entry.",
+                    "[ErrorCode:InvalidArgument] Config key is empty");
 
                 // SessionOptions.RegisterOrtExtensions can be manually tested by referencing the
                 // Microsoft.ML.OnnxRuntime.Extensions nuget package. After that is done, this should not throw.
-                ex = Assert.Throws<OnnxRuntimeException>(() => { opt.RegisterOrtExtensions(); });
-                Assert.Contains("Microsoft.ML.OnnxRuntime.Extensions NuGet package must be referenced", ex.Message);
-
+                AssertUtils.AssertThrowsCheckException<OnnxRuntimeException>(
+                    () => { opt.RegisterOrtExtensions(); },
+                    "RegisterOrtExtensions should throw if the Extensions package is not referenced",
+                    "Microsoft.ML.OnnxRuntime.Extensions NuGet package must be referenced");
+
+                // The below tests what happens when various execution providers are added
+                // to the session options. Due to changes in how execution providers are handled,
+                // only some are expected to throw exceptions.
+
+                // We can only check what EP's the package was built with for the
+                // Microsoft.ML.OnnxRuntime.Managed package because those contain the
+                // C# preprocessor directives (such as USE_CUDA). The Microsoft.ML.OnnxRuntime
+                // package will use the appropriate platform bindings (ie the Android bindings)
+                // where using the correct ifdefs for the C# preprocessor directives are
+                // impossible, so we use IfThrowsCheckException instead of using ifdefs.
```

Comment:
does this test run for both Microsoft.ML.OnnxRuntime and Microsoft.ML.OnnxRuntime.Managed?

Reply:
Yes, my understanding is that the Microsoft.ML.OnnxRuntime.Tests.Common are used to test both the C# bindings and the native NuGet package

Reply:
Microsoft.ML.OnnxRuntime.Tests.Common is C# code that only tests the C# bindings. 

The path from C# code to native code is in \csharp\src\Microsoft.ML.OnnxRuntime\NativeMethods.shared.cs where C# delegates are created to call functions from the C API.

The connection between this test code and the native code regarding the USE_XXX values happens when cmake runs a local build of the native code with `--build_csharp`. That's where cmake generates \csharp\Directory.Build.props which defines the USE_XXX EP values to match the native build as well as the path to the native onnxruntime library in the build output directory.

Reply:
Ah but these tests are run in the BrowserStack on-device tests as well, where the native Android bindings are tested.